kafka/consumer.go

```diff
@@ -142,16 +168,34 @@ func (c *Consumer) Run(ctx context.Context) error {
 	}
 }
 
+// fetch polls the Kafka broker for new records up to cfg.MaxPollRecords.
+// Any errors returned by fetch should be considered fatal and
 func (c *Consumer) fetch(ctx context.Context) error {
 	// NOTE(marclop) this is pretty naive consuming, to maximize throughput,
 	// it's best to use one goroutine per partition, but that requires more
 	// state management and blocking when rebalances happen.
 	c.mu.RLock()
 	defer c.mu.RUnlock()
-	fetches := c.client.PollFetches(ctx)
+
+	fetches := c.client.PollRecords(ctx, c.cfg.MaxPollRecords)
+	defer c.client.AllowRebalance()
```

Comment:
Should we be allowing rebalance immediately after committing offsets?

Reply:
I suppose we could since we've marked the records as committed. I was trying to avoid calling `c.client.AllowRebalance()` in different places, I can make so if you think it's best:

```go
fetches := c.client.PollRecords(ctx, c.cfg.MaxPollRecords)
if fetches.IsClientClosed() || errors.Is(fetches.Err0(), context.Canceled) {
	return context.Canceled // Client closed or context cancelled.
}
switch c.cfg.Delivery {
case apmqueue.AtLeastOnceDeliveryType:
	// Commit the fetched record offsets after we've processed them.
	defer func() {
		c.client.CommitUncommittedOffsets(ctx)
		c.client.AllowRebalance()
	}()
case apmqueue.AtMostOnceDeliveryType:
	// Commit the fetched record offsets as soon as we've polled them.
	err := c.client.CommitUncommittedOffsets(ctx)
	c.client.AllowRebalance()
	if err != nil {
		// If the commit fails, then return immediately and any uncommitted
		// records will be re-delivered in time. Otherwise, records may be
		// processed twice.
		return nil
	}
}
```

I think using `BlockRebalanceOnPoll` is still advantageous to try and reduce duplicates for `AtLeastOnceDeliveryType`, since it guards against that scenario.

Reply:
IIANM `AllowRebalance` is idempotent, so how about this?

```go
fetches := c.client.PollRecords(ctx, c.cfg.MaxPollRecords)
defer c.client.AllowRebalance()
...
switch c.cfg.Delivery {
case apmqueue.AtLeastOnceDeliveryType:
	// Commit the fetched record offsets after we've processed them.
	defer c.client.CommitUncommittedOffsets(ctx)
case apmqueue.AtMostOnceDeliveryType:
	// Commit the fetched record offsets as soon as we've polled them.
	if err := c.client.CommitUncommittedOffsets(ctx); err != nil {
		// If the commit fails, then return immediately and any uncommitted
		// records will be re-delivered in time. Otherwise, records may be
		// processed twice.
		return nil
	}
	// Allow rebalancing now that we have committed offsets, preventing
	// another consumer from reprocessing the records.
	c.client.AllowRebalance()
}
...
```

Reply:
LGTM. Thanks!